mitiq/rem/rem.py

```diff
@@ -0,0 +1,182 @@
+# Copyright (C) 2021 Unitary Fund
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Readout Confusion Inversion."""
+
+from typing import Callable, Optional, Union
+
+from functools import wraps
+import numpy as np
+
+from mitiq._typing import QPROGRAM, MeasurementResult
+from mitiq.executor.executor import Executor
+from mitiq.observable.observable import Observable
+
+
+def to_probability_vector(measurement: np.ndarray) -> np.ndarray:
+    """Convert a raw measurement to a probability vector.
+
+    Args:
+        measurement: A single measurement.
+    """
+    index = int("".join([str(m) for m in measurement]), base=2)
+    pv = np.zeros((2 ** measurement.shape[0]), dtype=np.uint8)
+    pv[index] = 1
+    return pv
+
+
+def sample_probability_vector(probability_vector: np.ndarray) -> np.ndarray:
+    """Sample a probability vector and return a raw measurement.
+
+    Args:
+        probability_vector: A probability vector.
+    """
+    result = np.random.choice(
+        len(probability_vector), size=1, p=probability_vector
+    )[0]
+    value = [int(i) for i in bin(result)[2:]]
+    return np.array(value, dtype=np.uint8)
+
+
+def mitigate_measurements(
+    noisy_result: MeasurementResult,
+    inverse_confusion_matrix: np.ndarray,
+) -> MeasurementResult:
+    """Applies the inverse confusion matrix against the noisy measurement
+    result and returns the adjusted measurements.
+
+    Args:
+        noisy_results: The unmitigated ``MeasurementResult``.
+        inverse_confusion_matrix: The inverse confusion matrix to apply to the
+            probability vector estimated with noisy measurement results.
+    """
+    assert noisy_result.qubit_indices is not None
+    num_qubits = len(noisy_result.qubit_indices)
+    assert inverse_confusion_matrix.shape == (2**num_qubits, 2**num_qubits)
+
+    empirical_prob_dist = np.apply_along_axis(
+        to_probability_vector, 1, noisy_result.asarray
+    )
+
+    adjusted_prob_dist = (inverse_confusion_matrix @ empirical_prob_dist.T).T
+
+    adjusted_result = np.apply_along_axis(
+        sample_probability_vector, 1, adjusted_prob_dist
+    )
```

Comment:
Here, you are basically applying REM to each individual measurement.
This is however problematic since, for a single measurement, the adjusted_prob_dist will typically contain negative elements. 
If I am not wrong, the correct way of applying REM is:

-  get a **single** empirical probability distribution associated to all the noisy measurements. 
- Multiply the inverse confusion matrix.
- Hopefully the result doesn't contain negative elements. If it does, we need to think about a way to handle this case.
- Resample new bitstrings from the adjusted distribution. The number of resampled bitstrings should equal to the number of the original noisy bistrings. 
- [Possible alternative to the previous point] Deduce a list of new bitstrings whose frequency best approximate the adjusted distribution.

Reply:
Can you share more about where we get negative elements?

Reply:
**The issue**
Even if the confusion matrix contains positive elements, its inverse (inverse confusion matrix) $M^{-1}$ typically contains negative elements. This means that a positive empirical distribution $p$, can be mapped to an adjusted distribution $p' = M^{-1} p$, where some elements of $p'$ are negative. 

Typically negative terms are very small since, by hypothesis, $p$ is not an arbitrary distribution but a noisy distribution having approximately the structure $p=M p_{\rm ideal}$. Unfortunately, negative elements are still possible due to statistical fluctuations or noise miscalibrations.

This effect is much stronger if REM is applied to individual shots (this is the current approach if I am not wrong), instead of applying REM to the averaged empirical distribution (i.e. the vector of frequencies of measured bitstrings).

You can reproduce this _negative probabilities_ issue, if you replace `p0=1` in your tests with something more intermediate e.g.  `p0=0.1`.

**Possible solution**
A possible solution is to clip small negative values of $p'$ to zero and renormalize. The pseudo code I have in mind is:
```python
# Clip negative numbers
adjusted_prob = np.array(adjusted_prob).clip(min=0)
# Renormalize
adjusted_prob = adjusted_prob / np.sum(adjusted_prob)
```

Reply:
Thanks for this explanation. I tried with p0=0.1 and I can see now what you were concerned about. I'll need a little more time with this to correct it.

One question for now:
> get a single empirical probability distribution associated to all the noisy measurements.

I get that this will be a single vector corresponding to the possible outcomes (e.g. measurements for 2 qubits would have four values corresponding to 00, 01, 10, 11). However, should these be raw counts or should they sum to 1?

`[128, 256, 512, 128]` vs. `[0.125, 0.25, 0.5, 0.125]`

Reply:
> Thanks for this explanation. I tried with p0=0.1 and I can see now what you were concerned about. I'll need a little more time with this to correct it.
> 
> One question for now:
> 
> > get a single empirical probability distribution associated to all the noisy measurements.
> 
> I get that this will be a single vector corresponding to the possible outcomes (e.g. measurements for 2 qubits would have four values corresponding to 00, 01, 10, 11). However, should these be raw counts or should they sum to 1?
> 
> `[128, 256, 512, 128]` vs. `[0.125, 0.25, 0.5, 0.125]`

I would maybe normalize such that sum=1. But I don't have a strong preference.  Up to you. 

Note that even if using integer counts (without normalization), after multiplying for $M^{-1}$ the result is a vector with non-integer elements. So I don't see a strong advantage in using integer counts compared to normalized distributions.

Reply:
Please have a look. I clip the negative values as you suggested.